### PR TITLE
Code review

### PR DESCRIPTION
**Pull request by neworderofjamie**
All of your feedback on the Python code review implemented aside from some stuff relating to neuron parameters in general which needs a little more thought. Would appreciate a brief check from @mundya and @mossblaser. Also, I have yet to touch the spinnaker parameters dictionary - requires a little more thought.

### CODE REVIEW

File: simulator.py [hunk 1]
```diff
@@ -318,18 +334,18 @@ def _load_current_input_verts(self, placements, allocations,
...
             # Build direct connection for projection
-            direct_weights = proj.build_direct_connection()
+            direct_weights = proj._build_direct_connection()
 
             # Loop through synapse verts
             for v in c_cluster.verts:
-                logger.debug("\t\tVertex %s" % v)
+                logger.debug("\t\tVertex %s", v)
 
                 # Use native S15.16 format
```
Comment (mundya) [inline simulator.py]: Adjust the comment as well?

File: simulator.py [hunk 2]
```diff
@@ -39,6 +39,7 @@ def __init__(self, n):
 class State(common.control.BaseState):
     def __init__(self):
         common.control.BaseState.__init__(self)
+        # JK: Can these be removed
         self.mpi_rank = 0
```
Comment (mossblaser) [inline simulator.py]: Gross! I hope so! ;)

File: simulator.py [hunk 3]
```diff
@@ -464,29 +480,16 @@ def _build(self, duration_ms):
...
-
-            # Reserve these extra cores (above monitor) on each chip
-            reservation = slice(1, 1 + self.reserve_extra_cores_per_chip)
-            constraints.append(ReserveResourceConstraint(machine.Cores,
-                                                         reservation))
-
-        logger.info("Placing and routing")
+        # Get system info
+        system_info = self.machine_controller.get_system_info()
+        logger.debug("Found %ux%u chip machine",
```
Comment (mossblaser) [inline simulator.py]: May be more truthful to print len(system_info) giving the actual number of chips (though perhaps I'm not realising what you're using that information for?)

File: spinnaker/current_input_cluster.py [hunk 4]
```diff
@@ -21,12 +21,12 @@
 # ------------------------------------------------------------------------------
 class Regions(enum.IntEnum):
     """Region names, corresponding to those defined in `ensemble.h`"""
```
Comment (mossblaser) [inline spinnaker/current_input_cluster.py]: `ensemble.h`

File: simulator.py (hunk 3 above)
Comment (neworderofjamie) [inline simulator.py]: Yeah, in the case of oddly shaped spalloc'd machines, that would probably make more sense - while this is purely diagnostic information, I think that number is always going to be more useful

File: spinnaker/synapse_cluster.py [hunk 5]
```diff
@@ -21,12 +21,12 @@
 # ------------------------------------------------------------------------------
 class Regions(enum.IntEnum):
     """Region names, corresponding to those defined in `ensemble.h`"""
```
Comment (mossblaser) [inline spinnaker/synapse_cluster.py]: `ensemble.h`

**Comment by mossblaser** [comment]
I'm afraid nothing much more than a brief glance from me but looks sensible to my inexpert eye!

**Comment by mundya** [comment]
> I'm afraid nothing much more than a brief glance from me but looks sensible to my inexpert eye!

Likewise

**Comment by neworderofjamie** [comment]
Great! after our marathon in-person code review, that's all that I think's
required - thanks guys

**Pull request merged**